### ISSUE TITLE
new feature sycl::complex

### ISSUE THREAD

**Issue by bd4** [AdaptiveCpp/AdaptiveCpp#341]
Re the inconsistency of std::complex working in device kernels (see #340), I think it's worthwhile to implement a sycl::complex type and propose it for standardization.

This can be done in two phases:

1. data structure, basic arithmetic operations, norm, conj, abs
2. math lib functions (trig functions, pow, exp, log etc)

The first phase should be very simple. (2) is more work, and should leverage existing implementations as much as possible.

**Comment by illuhad** [comment]
Thank you! we might have to refactor the way math functions are formulated anyway, so 2 might indeed be a larger step. Let's start with 1 - let me know if you need guidance or have questions if you start working on this.
If you haven't already done so, reading this might help:
https://github.com/illuhad/hipSYCL/blob/develop/doc/architecture.md
https://github.com/illuhad/hipSYCL/blob/develop/doc/macros.md

**Comment by bd4** [comment]
Some colleagues at ALCF have been working on this, initially targeting the Intel llvm based SYCL implementation: https://github.com/argonne-lcf/SyclCPLX. There is working getting this into the standard, but in the meantime, this is a single header implementation, using `sycl::ext::cplx` namespace rather than just `sycl`.

When I try to build the example using hipsycl (devlopment branch, omp backend), I get errors that `isinf`, `isnan` and `half` are not in the `sycl` namespace. Are there plans to add these to hipSYCL?

**Comment by illuhad** [comment]
`isnan`  seems to be implemented: https://github.com/illuhad/hipSYCL/blob/476b6f21346e8fb14cc19904e04516a5df0b183a/include/hipSYCL/sycl/libkernel/builtins.hpp#L965

I think `isinf` (like most of the relational builtins) is still missing, but it should be easy to add.

`half` is an optional feature in SYCL, and is therefore currently not a priority. The type should exist though; we just map it to an `int16`, so any float operation on it will not work: https://github.com/illuhad/hipSYCL/blob/8294ac792301d0d62f9ac83a063cfcc169c1d441/include/hipSYCL/sycl/types.hpp#L106

**Comment by bd4** [comment]
Thanks! Looking at the standard, there is a way to query a device for half support, via `aspect::fp16`, but I don't see anything re a global flag indicating in the runtime if it's not supported at all. From here:
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_scalar_data_types

It seems to imply the type should always be defined even if no devices support it, but it's not totally clear.

For the purpose of getting SyclCPLX to work with hipSYCL in the short term, we could add some ifdef around the sycl::half implementations. I'm not sure what exactly to check for though, some ugly hardcoding is probably need, e.g. check for `__INTEL_CLANG_COMPILER` and enable it.

**Comment by illuhad** [comment]
That's right, according to the spec, kernels should still compile even if optional device features are not supported. The idea is that this would then lead to an error at runtime when the kernel is actually invoked on such a device. However this is something that cannot be supported by all implementations (e.g. library-only implementations).

As I've said, we *do* have the type defined.

The spec is very vague what this `half` should actually be able to do. There's a lot of uncertainty here.

From the hipSYCL perspective, one major challenge of implementing it is that we support operating as library for non-clang compilers, in which case we can assume neither clang's `__fp16`nor `_Fp16`, clang targets where `__fp16` is supported but not `_Fp16`, and targets where both `__fp16`and `_Fp16` are supported. Half support in current C++ compilers is a mess, and it's quite some effort to find abstractions that work efficiently for all those cases.

hipSYCL defines `__HIPSYCL__` which can also be used to detect it using macros.

**Comment by bd4** [comment]
Looks like sycl::half is not an issue with the simple hello_world example at least. Here are the changes I needed to get it to build with the omp backend:
https://github.com/argonne-lcf/SyclCPLX/pull/21

I build like this:
```
syclcc -fsycl -O2 --hipsycl-targets='omp' -I./include -o hello_world hello_world.cpp
```

Getting it to build with device backends is more involved, but this is a start.

**Comment by bd4** [comment]
For SyclCPLX, we were able to fix the namespace issue, and now what remains is that the following functions are not defined in hipsycl namespace (for real numbers): `isinf`, `isfinite`, `ldexp`, and `signbit`. I am not sure I am following all the macro magic, but looks like hipSYCL defines host versions by calling out to std::X, and device versions with `__spirv_ocl_X`, and an interface in libkernel/builtins.hpp. Looks like it wouldn't be too hard to add four more.

**Comment by illuhad** [comment]
Thanks for the feedback! Probably the main reason why those are not implemented is because I didn't want anymore after the hundreds of other builtins ;) There are some that are more difficult because their interface is different and does not fit well into the mechanisms that we have, but I don't think these ones are affected.

It's not super difficult, but slightly more effort than you describe: `__spirv_ocl_` is only for the SPIR-V library backend. There are also implementations in CUDA, HIP, and for host and with #862, there are additional backend-specific bitcode libraries that need to be implemented with support for new builtins.

**Comment by nmnobre** [comment]
> For SyclCPLX, we were able to fix the namespace issue, and now what remains is that the following functions are not defined in hipsycl namespace (for real numbers): `isinf`, `isfinite`, `ldexp`, and `signbit`. I am not sure I am following all the macro magic, but looks like hipSYCL defines host versions by calling out to std::X, and device versions with `__spirv_ocl_X`, and an interface in libkernel/builtins.hpp. Looks like it wouldn't be too hard to add four more.

Hi @bd4, could you try merging both #959 and #962 and let us know if those work for your workflow?

**Comment by bd4** [comment]
@nmnobre thank you for your work on this! So far I can now compile and run a simple SyclCPLX example using the omp backend, with a change to call in the `sycl_ext_cplx.hpp` header `sycl::atan2` call by casting it's args rather than doing `sycl::atan2<ValueType>(...)`. Not sure if this is our implementation relying on templating not specified by the standard or not, in any case is an easy fix. I will continue to test more thoroughly, but so far it seems the newly implemented math functions are getting us much closer.

Edit: for reference, here is my branch with those two PRs merged: https://github.com/bd4/OpenSYCL/tree/new-math-fn

**Comment by nmnobre** [comment]
I see, `sycl::atan2` is indeed not a template and perhaps you're calling it with integral arguments?
In that case the compiler will complain about the ambiguity between the available `float` and `double` implementations, so a cast will indeed be needed. I'm glad you are getting closer :)
